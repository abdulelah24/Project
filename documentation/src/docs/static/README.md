# JUnit 5 User Guide

This branch contains the website sources available at:
http://junit-team.github.io/junit-lambda/
