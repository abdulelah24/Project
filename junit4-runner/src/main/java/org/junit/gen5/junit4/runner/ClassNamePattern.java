/*
 * Copyright 2015-2016 the original author or authors.
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution and is available at
 *
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.junit.gen5.junit4.runner;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Inherited;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * {@code @ClassNamePattern} specifies a regular expression that is used
 * to match against class names when a test class annotated with
 * {@code @RunWith(JUnit5.class)} is executed.
 *
 * @since 5.0
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
@Inherited
@Documented
public @interface ClassNamePattern {

	/**
	 * Regular expression used to match against class names.
	 */
	String value();

}
