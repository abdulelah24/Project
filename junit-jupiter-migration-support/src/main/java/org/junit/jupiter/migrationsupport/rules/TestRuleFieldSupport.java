/*
 * Copyright 2015-2016 the original author or authors.
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution and is available at
 *
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.junit.jupiter.migrationsupport.rules;

import static org.junit.platform.commons.meta.API.Usage.Experimental;

import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.lang.reflect.Member;
import java.util.Arrays;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.junit.jupiter.api.extension.TestExtensionContext;
import org.junit.jupiter.migrationsupport.rules.adapter.AbstractTestRuleAdapter;
import org.junit.jupiter.migrationsupport.rules.member.RuleAnnotatedField;
import org.junit.jupiter.migrationsupport.rules.member.RuleAnnotatedMember;
import org.junit.platform.commons.meta.API;
import org.junit.rules.TestRule;

@API(Experimental)
class TestRuleFieldSupport extends AbstractTestRuleSupport {

	TestRuleFieldSupport(Function<RuleAnnotatedMember, AbstractTestRuleAdapter> adapterGenerator,
			Class<? extends TestRule> ruleType) {
		super(adapterGenerator, ruleType);
	}

	@Override
	protected RuleAnnotatedMember createRuleAnnotatedMember(TestExtensionContext context, Member member) {
		return new RuleAnnotatedField(context.getTestInstance(), (Field) member);
	}

	@Override
	protected List<Member> findRuleAnnotatedMembers(Object testInstance) {
		return findAnnotatedFields(testInstance, super.getRuleType(), super.getAnnotationType()).collect(
			Collectors.toList());
	}

	// TODO: decide whether this should be promoted to AnnotationUtils
	private static Stream<Field> findAnnotatedFields(Object instance, Class<?> fieldType,
			Class<? extends Annotation> annotationType) {
		Field[] fields = instance.getClass().getFields();

		// @formatter:off
        return Arrays.stream(fields)
                .filter(field -> fieldType.isAssignableFrom(field.getType()))
                .filter(field -> field.isAnnotationPresent(annotationType));
		// @formatter:on
	}

}
