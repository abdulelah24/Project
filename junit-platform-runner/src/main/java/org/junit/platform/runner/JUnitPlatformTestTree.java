/*
 * Copyright 2015-2017 the original author or authors.
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v2.0 which
 * accompanies this distribution and is available at
 *
 * http://www.eclipse.org/legal/epl-v20.html
 */

package org.junit.platform.runner;

import static java.util.stream.Collectors.toCollection;
import static java.util.stream.Collectors.toSet;

import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import java.util.function.Predicate;

import org.junit.platform.commons.util.AnnotationUtils;
import org.junit.platform.commons.util.Preconditions;
import org.junit.platform.commons.util.StringUtils;
import org.junit.platform.engine.FilterResult;
import org.junit.platform.engine.TestDescriptor;
import org.junit.platform.engine.TestSource;
import org.junit.platform.engine.support.descriptor.ClassSource;
import org.junit.platform.engine.support.descriptor.MethodSource;
import org.junit.platform.launcher.PostDiscoveryFilter;
import org.junit.platform.launcher.TestCollection;
import org.junit.platform.launcher.TestIdentifier;
import org.junit.platform.launcher.TestPlan;
import org.junit.platform.suite.api.SuiteDisplayName;
import org.junit.platform.suite.api.UseTechnicalNames;
import org.junit.runner.Description;
import org.junit.runner.manipulation.Filter;
import org.junit.runner.manipulation.NoTestsRemainException;
import org.junit.runner.notification.RunNotifier;

/**
 * @since 1.0
 */
class JUnitPlatformTestTree {

	private final Map<TestIdentifier, Description> descriptions = new HashMap<>();
	private final TestCollection testCollection;
	private final Function<TestIdentifier, String> nameExtractor;
	private final String suiteDisplayName;
	private Description suiteDescription;

	JUnitPlatformTestTree(TestCollection testCollection, Class<?> testClass) {
		this.testCollection = testCollection;
		if (useTechnicalNames(testClass)) {
			this.nameExtractor = JUnitPlatformTestTree::getTechnicalName;
			this.suiteDisplayName = testClass.getName();
		}
		else {
			this.nameExtractor = TestIdentifier::getDisplayName;
			this.suiteDisplayName = getSuiteDisplayName(testClass);
		}
		this.suiteDescription = generateSuiteDescription(testCollection);
	}

	private static boolean useTechnicalNames(Class<?> testClass) {
		return testClass.isAnnotationPresent(UseTechnicalNames.class);
	}

	Description getSuiteDescription() {
		return this.suiteDescription;
	}

	Description getDescription(TestIdentifier identifier) {
		return this.descriptions.get(identifier);
	}

	void executeTests(RunNotifier notifier) {
		JUnitPlatformRunnerListener listener = new JUnitPlatformRunnerListener(this, notifier);
		this.testCollection.execute(listener);
	}

	void filter(Filter filter) throws NoTestsRemainException {
		Set<TestIdentifier> filteredIdentifiers = getFilteredLeaves(filter);
		if (filteredIdentifiers.isEmpty()) {
			throw new NoTestsRemainException();
		}

		PostDiscoveryFilter postDiscoveryFilter = (TestDescriptor t) -> {
			return FilterResult.includedIf(filteredIdentifiers.contains(TestIdentifier.from(t)));
		};
		testCollection.applyPostDiscoveryFilters(postDiscoveryFilter);

		// Update our fields to match the testCollection
		descriptions.clear();
		suiteDescription = generateSuiteDescription(testCollection);
	}

	private Description generateSuiteDescription(TestCollection testCollection) {
		Description suiteDescription = Description.createSuiteDescription(suiteDisplayName);
		buildDescriptionTree(suiteDescription, testCollection);
		return suiteDescription;
	}

	private String getSuiteDisplayName(Class<?> testClass) {
		// @formatter:off
		return AnnotationUtils.findAnnotation(testClass, SuiteDisplayName.class)
				.map(SuiteDisplayName::value)
				.filter(StringUtils::isNotBlank)
				.orElse(testClass.getName());
		// @formatter:on
	}

	private void buildDescriptionTree(Description suiteDescription, TestCollection testCollection) {
		TestPlan testPlan = testCollection.testPlan();
		testPlan.getRoots().stream().filter(testIdentifier -> !testPlan.getChildren(testIdentifier).isEmpty()).forEach(
			testIdentifier -> buildDescription(testIdentifier, suiteDescription, testPlan));
	}

	void addDynamicDescription(TestIdentifier newIdentifier, String parentId) {
		TestPlan plan = this.testCollection.testPlan();
		TestIdentifier parentTestIdentifier = plan.getTestIdentifier(parentId);
		Description parent = getDescription(parentTestIdentifier);

		if (parent == null) {
			boolean parentIsRoot = plan.getRoots().contains(parentTestIdentifier);
			Preconditions.condition(parentIsRoot,
				() -> String.format("No matching Description for parentId=%s (testIdentifer=%s)", parentId,
					parentTestIdentifier));
			// This root was filtered out in buildDescriptionTree() because it had no children; add it back
			parent = buildDescription(parentTestIdentifier, suiteDescription, plan);
		}

		plan.add(newIdentifier);
		buildDescription(newIdentifier, parent, plan);
	}

	private Description buildDescription(TestIdentifier identifier, Description parent, TestPlan testPlan) {
		Description newDescription = createJUnit4Description(identifier, testPlan);
		parent.addChild(newDescription);
		this.descriptions.put(identifier, newDescription);
		testPlan.getChildren(identifier).forEach(
			testIdentifier -> buildDescription(testIdentifier, newDescription, testPlan));
		return newDescription;
	}

	private Description createJUnit4Description(TestIdentifier identifier, TestPlan testPlan) {
		String name = nameExtractor.apply(identifier);
		if (identifier.isTest()) {
			String containerName = testPlan.getParent(identifier).map(nameExtractor).orElse("<unrooted>");
			return Description.createTestDescription(containerName, name, identifier.getUniqueId());
		}
		return Description.createSuiteDescription(name, identifier.getUniqueId());
	}

	private static String getTechnicalName(TestIdentifier testIdentifier) {
		Optional<TestSource> optionalSource = testIdentifier.getSource();
		if (optionalSource.isPresent()) {
			TestSource source = optionalSource.get();
			if (source instanceof ClassSource) {
				return ((ClassSource) source).getJavaClass().getName();
			}
			else if (source instanceof MethodSource) {
				MethodSource methodSource = (MethodSource) source;
				String methodParameterTypes = methodSource.getMethodParameterTypes();
				if (StringUtils.isBlank(methodParameterTypes)) {
					return methodSource.getMethodName();
				}
				return String.format("%s(%s)", methodSource.getMethodName(), methodParameterTypes);
			}
		}

		// Else fall back to display name
		return testIdentifier.getDisplayName();
	}

	Set<TestIdentifier> getTestsInSubtree(TestIdentifier ancestor) {
		// @formatter:off
		return testCollection.testPlan().getDescendants(ancestor).stream()
				.filter(TestIdentifier::isTest)
				.collect(toCollection(LinkedHashSet::new));
		// @formatter:on
	}

	private Set<TestIdentifier> getFilteredLeaves(Filter filter) {
		Set<TestIdentifier> identifiers = applyFilterToDescriptions(filter);
		return removeNonLeafIdentifiers(identifiers);
	}

	private Set<TestIdentifier> removeNonLeafIdentifiers(Set<TestIdentifier> identifiers) {
		return identifiers.stream().filter(isALeaf(identifiers)).collect(toSet());
	}

	private Predicate<? super TestIdentifier> isALeaf(Set<TestIdentifier> identifiers) {
		return testIdentifier -> {
			TestPlan plan = testCollection.testPlan();
			Set<TestIdentifier> descendants = plan.getDescendants(testIdentifier);
			return identifiers.stream().noneMatch(descendants::contains);
		};
	}

	private Set<TestIdentifier> applyFilterToDescriptions(Filter filter) {
		// @formatter:off
		return descriptions.entrySet()
				.stream()
				.filter(entry -> filter.shouldRun(entry.getValue()))
				.map(Entry::getKey)
				.collect(toSet());
		// @formatter:on
	}

}
