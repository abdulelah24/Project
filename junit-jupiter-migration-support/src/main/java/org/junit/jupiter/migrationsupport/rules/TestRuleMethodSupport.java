/*
 * Copyright 2015-2016 the original author or authors.
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution and is available at
 *
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.junit.jupiter.migrationsupport.rules;

import static org.junit.platform.commons.meta.API.Usage.Experimental;
import static org.junit.platform.commons.util.ReflectionUtils.MethodSortOrder.HierarchyDown;

import java.lang.reflect.Member;
import java.lang.reflect.Method;
import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import org.junit.jupiter.api.extension.TestExtensionContext;
import org.junit.jupiter.migrationsupport.rules.adapter.AbstractTestRuleAdapter;
import org.junit.jupiter.migrationsupport.rules.member.RuleAnnotatedMember;
import org.junit.jupiter.migrationsupport.rules.member.RuleAnnotatedMethod;
import org.junit.platform.commons.meta.API;
import org.junit.platform.commons.util.AnnotationUtils;
import org.junit.rules.TestRule;

@API(Experimental)
class TestRuleMethodSupport extends AbstractTestRuleSupport {

	TestRuleMethodSupport(Function<RuleAnnotatedMember, AbstractTestRuleAdapter> adapterGenerator,
			Class<? extends TestRule> ruleType) {
		super(adapterGenerator, ruleType);
	}

	@Override
	protected RuleAnnotatedMember createRuleAnnotatedMember(TestExtensionContext context, Member member) {
		return new RuleAnnotatedMethod(context.getTestInstance(), (Method) member);
	}

	@Override
	protected List<Member> findRuleAnnotatedMembers(Object testInstance) {
		List<Method> annotatedMethods = AnnotationUtils.findAnnotatedMethods(testInstance.getClass(),
			super.getAnnotationType(), HierarchyDown);

		Predicate<Method> methodsWithCorrectReturnType = method -> method.getReturnType().isAssignableFrom(
			super.getRuleType());

		// @formatter:off
        return annotatedMethods.stream()
                .filter(methodsWithCorrectReturnType)
                .collect(Collectors.toList());
		// @formatter:on

	}

}
