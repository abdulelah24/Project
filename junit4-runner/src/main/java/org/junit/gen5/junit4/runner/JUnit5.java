/*
 * Copyright 2015-2016 the original author or authors.
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution and is available at
 *
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.junit.gen5.junit4.runner;

import static java.util.Arrays.stream;
import static java.util.stream.Collectors.toList;
import static org.junit.gen5.engine.specification.dsl.ClassFilters.classNameMatches;
import static org.junit.gen5.engine.specification.dsl.ClassSelectorBuilder.forClass;
import static org.junit.gen5.engine.specification.dsl.DiscoveryRequestBuilder.request;

import java.lang.annotation.Annotation;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.function.Function;

import org.junit.gen5.commons.util.Preconditions;
import org.junit.gen5.engine.DiscoveryRequest;
import org.junit.gen5.engine.DiscoverySelector;
import org.junit.gen5.engine.PostDiscoveryFilter;
import org.junit.gen5.engine.specification.dsl.*;
import org.junit.gen5.launcher.Launcher;
import org.junit.gen5.launcher.TestIdentifier;
import org.junit.gen5.launcher.TestPlan;
import org.junit.runner.Description;
import org.junit.runner.Runner;
import org.junit.runner.manipulation.Filter;
import org.junit.runner.manipulation.Filterable;
import org.junit.runner.manipulation.NoTestsRemainException;
import org.junit.runner.notification.RunNotifier;
import org.junit.runners.model.InitializationError;

/**
 * JUnit 4 based {@link Runner} which runs tests that use the JUnit 5 programming and extension models.
 *
 * <p>Annotating a class with {@code @RunWith(JUnit5.class)} allows it to be run with IDEs and build systems that
 * support JUnit 4 but do not yet support the JUnit 5 APIs directly.
 *
 * <p>Consult the various annotations in this package for configuration options.
 *
 * <p>If you don't use any annotations, you can simply us this runner on a JUnit 5 test class.
 * Contrary to standard JUnit 5 test classes, the test class must be {@code public} in order
 * to be picked up by IDEs and build tools.
 *
 * @since 5.0
 * @see Classes
 * @see ClassNamePattern
 * @see Packages
 * @see UniqueIds
 * @see OnlyIncludeTags
 * @see ExcludeTags
 * @see OnlyEngine
 */
public class JUnit5 extends Runner implements Filterable {

	private static final Class<?>[] EMPTY_CLASS_ARRAY = new Class<?>[0];
	private static final String[] EMPTY_STRING_ARRAY = new String[0];
	private static final String EMPTY_STRING = "";

	private final Launcher launcher = new Launcher();
	private final Class<?> testClass;
	private DiscoveryRequest specification;
	private JUnit5TestTree testTree;

	public JUnit5(Class<?> testClass) throws InitializationError {
		this.testClass = testClass;
		this.specification = createSpecification();
		this.testTree = generateTestTree();
	}

	@Override
	public Description getDescription() {
		return this.testTree.getSuiteDescription();
	}

	@Override
	public void run(RunNotifier notifier) {
		JUnit5RunnerListener listener = new JUnit5RunnerListener(this.testTree, notifier);
		this.launcher.registerTestExecutionListeners(listener);
		this.launcher.execute(this.specification);
	}

	private JUnit5TestTree generateTestTree() {
		Preconditions.notNull(this.specification, "DiscoveryRequest must not be null");
		TestPlan plan = this.launcher.discover(this.specification);
		return new JUnit5TestTree(plan, testClass);
	}

	private DiscoveryRequest createSpecification() {
		List<DiscoverySelector> selectors = getSpecElementsFromAnnotations();

		// Allows to simply add @RunWith(JUnit5.class) to any JUnit5 test case
		if (selectors.isEmpty()) {
			selectors.add(forClass(this.testClass));
		}

		DiscoveryRequest request = request().select(selectors).build();
		addFiltersFromAnnotations(request);
		return request;
	}

	private void addFiltersFromAnnotations(DiscoveryRequest request) {
		addClassNameMatchesFilter(request);
		addIncludeTagsFilter(request);
		addExcludeTagsFilter(request);
	}

	private List<DiscoverySelector> getSpecElementsFromAnnotations() {
		List<DiscoverySelector> selectors = new ArrayList<>();
		selectors.addAll(getClassSpecificationElements());
		selectors.addAll(getUniqueIdSpecificationElements());
		selectors.addAll(getPackageSpecificationElements());
		return selectors;
	}

	private List<DiscoverySelector> getClassSpecificationElements() {
		return stream(getTestClasses()).map(ClassSelectorBuilder::forClass).collect(toList());
	}

	private List<DiscoverySelector> getUniqueIdSpecificationElements() {
		return stream(getUniqueIds()).map(UniqueIdSelectorBuilder::byUniqueId).collect(toList());
	}

	private List<DiscoverySelector> getPackageSpecificationElements() {
		return stream(getPackageNames()).map(PackageSelectorBuilder::byPackageName).collect(toList());
	}

	private void addClassNameMatchesFilter(DiscoveryRequest plan) {
		String regex = getClassNameRegExPattern();
		if (!regex.isEmpty()) {
			plan.addFilter(classNameMatches(regex));
		}
	}

	private void addIncludeTagsFilter(DiscoveryRequest plan) {
		String[] includeTags = getIncludeTags();
		if (includeTags.length > 0) {
			PostDiscoveryFilter tagNamesFilter = TagFilterBuilder.includeTags(includeTags);
			plan.addPostFilter(tagNamesFilter);
		}
	}

	private void addExcludeTagsFilter(DiscoveryRequest plan) {
		String[] excludeTags = getExcludeTags();
		if (excludeTags.length > 0) {
			PostDiscoveryFilter excludeTagsFilter = TagFilterBuilder.excludeTags(excludeTags);
			plan.addPostFilter(excludeTagsFilter);
		}
	}

	private Class<?>[] getTestClasses() {
		return getValueFromAnnotation(Classes.class, Classes::value, EMPTY_CLASS_ARRAY);
	}

	private String[] getUniqueIds() {
		return getValueFromAnnotation(UniqueIds.class, UniqueIds::value, EMPTY_STRING_ARRAY);
	}

	private String[] getPackageNames() {
		return getValueFromAnnotation(Packages.class, Packages::value, EMPTY_STRING_ARRAY);
	}

	private String[] getIncludeTags() {
		return getValueFromAnnotation(OnlyIncludeTags.class, OnlyIncludeTags::value, EMPTY_STRING_ARRAY);
	}

	private String[] getExcludeTags() {
		return getValueFromAnnotation(ExcludeTags.class, ExcludeTags::value, EMPTY_STRING_ARRAY);
	}

	private String getExplicitEngineId() {
		return getValueFromAnnotation(OnlyEngine.class, OnlyEngine::value, EMPTY_STRING).trim();
	}

	private String getClassNameRegExPattern() {
		return getValueFromAnnotation(ClassNamePattern.class, ClassNamePattern::value, EMPTY_STRING).trim();
	}

	private <A extends Annotation, V> V getValueFromAnnotation(Class<A> annotationClass, Function<A, V> extractor,
			V defaultValue) {
		A annotation = this.testClass.getAnnotation(annotationClass);
		return (annotation != null ? extractor.apply(annotation) : defaultValue);
	}

	@Override
	public void filter(Filter filter) throws NoTestsRemainException {
		Set<TestIdentifier> filteredIdentifiers = testTree.getFilteredLeaves(filter);
		if (filteredIdentifiers.isEmpty()) {
			throw new NoTestsRemainException();
		}
		this.specification = createTestPlanSpecificationForUniqueIds(filteredIdentifiers);
		this.testTree = generateTestTree();
	}

	private DiscoveryRequest createTestPlanSpecificationForUniqueIds(Set<TestIdentifier> testIdentifiers) {
		// @formatter:off
		List<DiscoverySelector> elements = testIdentifiers.stream()
				.map(TestIdentifier::getUniqueId)
				.map(Object::toString)
				.map(UniqueIdSelectorBuilder::byUniqueId)
				.collect(toList());
		// @formatter:on
		return request().select(elements).build();
	}

}
