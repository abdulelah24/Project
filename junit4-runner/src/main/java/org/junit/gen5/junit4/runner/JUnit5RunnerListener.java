/*
 * Copyright 2015-2016 the original author or authors.
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution and is available at
 *
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.junit.gen5.junit4.runner;

import static org.junit.gen5.engine.TestExecutionResult.Status.*;

import org.junit.gen5.engine.TestExecutionResult;
import org.junit.gen5.engine.TestExecutionResult.Status;
import org.junit.gen5.launcher.TestExecutionListener;
import org.junit.gen5.launcher.TestIdentifier;
import org.junit.runner.Description;
import org.junit.runner.notification.Failure;
import org.junit.runner.notification.RunNotifier;

/**
 * @since 5.0
 */
class JUnit5RunnerListener implements TestExecutionListener {

	private final JUnit5TestTree testTree;
	private final RunNotifier notifier;

	JUnit5RunnerListener(JUnit5TestTree testTree, RunNotifier notifier) {
		this.testTree = testTree;
		this.notifier = notifier;
	}

	@Override
	public void dynamicTestRegistered(TestIdentifier testIdentifier) {
		System.err.println("JUnit5 Runner does not support dynamic tests.");
	}

	@Override
	public void executionSkipped(TestIdentifier testIdentifier, String reason) {
		Description description = findJUnit4Description(testIdentifier);
		this.notifier.fireTestIgnored(description);
	}

	@Override
	public void executionStarted(TestIdentifier testIdentifier) {
		if (testIdentifier.isTest()) {
			Description description = findJUnit4Description(testIdentifier);
			this.notifier.fireTestStarted(description);
		}
	}

	@Override
	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
		Description description = findJUnit4Description(testIdentifier);
		Status status = testExecutionResult.getStatus();
		if (status == ABORTED) {
			this.notifier.fireTestAssumptionFailed(toFailure(testExecutionResult, description));
		}
		else if (status == FAILED) {
			this.notifier.fireTestFailure(toFailure(testExecutionResult, description));
		}
		if (testIdentifier.isTest()) {
			this.notifier.fireTestFinished(description);
		}
	}

	private Failure toFailure(TestExecutionResult testExecutionResult, Description description) {
		return new Failure(description, testExecutionResult.getThrowable().orElse(null));
	}

	private Description findJUnit4Description(TestIdentifier testIdentifier) {
		return this.testTree.getDescription(testIdentifier);
	}

}
