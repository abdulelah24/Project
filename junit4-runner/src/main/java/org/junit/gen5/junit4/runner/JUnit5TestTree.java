/*
 * Copyright 2015-2016 the original author or authors.
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution and is available at
 *
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.junit.gen5.junit4.runner;

import java.util.HashMap;
import java.util.Map;

import org.junit.gen5.launcher.TestIdentifier;
import org.junit.gen5.launcher.TestPlan;
import org.junit.runner.Description;

/**
 * @since 5.0
 */
class JUnit5TestTree {

	private final Map<TestIdentifier, Description> descriptions = new HashMap<>();
	private final Description suiteDescription;

	JUnit5TestTree(TestPlan plan, Class<?> testClass) {
		this.suiteDescription = generateDescription(plan, testClass);
	}

	Description getSuiteDescription() {
		return this.suiteDescription;
	}

	Description getDescription(TestIdentifier identifier) {
		return this.descriptions.get(identifier);
	}

	private Description generateDescription(TestPlan testPlan, Class<?> testClass) {
		Description suiteDescription = Description.createSuiteDescription(testClass.getName());
		buildDescriptionTree(suiteDescription, testPlan);
		return suiteDescription;
	}

	private void buildDescriptionTree(Description suiteDescription, TestPlan testPlan) {
		testPlan.getRoots().stream().forEach(
			testIdentifier -> buildDescription(testIdentifier, suiteDescription, testPlan));
	}

	private void buildDescription(TestIdentifier identifier, Description parent, TestPlan testPlan) {
		Description newDescription = createJUnit4Description(identifier, testPlan);
		parent.addChild(newDescription);
		this.descriptions.put(identifier, newDescription);
		testPlan.getChildren(identifier).stream().forEach(
			testIdentifier -> buildDescription(testIdentifier, newDescription, testPlan));
	}

	private Description createJUnit4Description(TestIdentifier identifier, TestPlan testPlan) {
		if (identifier.isTest()) {
			return Description.createTestDescription(testPlan.getParent(identifier).get().getDisplayName(),
				identifier.getDisplayName(), identifier.getUniqueId());
		}
		else {
			return Description.createSuiteDescription(identifier.getDisplayName(), identifier.getUniqueId());
		}
	}

}
