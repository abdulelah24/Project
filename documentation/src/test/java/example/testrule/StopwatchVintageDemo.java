/*
 * Copyright 2015-2017 the original author or authors.
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution and is available at
 *
 * http://www.eclipse.org/legal/epl-v10.html
 */

package example.testrule;

// tag::user_guide[]
import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;
import static org.junit.Assume.assumeTrue;

import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;

import org.junit.AssumptionViolatedException;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.Stopwatch;
import org.junit.runner.Description;

public class StopwatchVintageDemo {

	private static final Logger logger = Logger.getLogger("");

	private static void logInfo(Description description, String status, long nanos) {
		String testName = description.getMethodName();
		logger.info(
			String.format("Test %s %s, spent %d microseconds", testName, status, TimeUnit.NANOSECONDS.toMicros(nanos)));
	}

	@Rule
	public Stopwatch stopwatch = new Stopwatch() {
		@Override
		protected void succeeded(long nanos, Description description) {
			logInfo(description, "succeeded", nanos);
		}

		@Override
		protected void failed(long nanos, Throwable e, Description description) {
			logInfo(description, "failed", nanos);
		}

		@Override
		protected void skipped(long nanos, AssumptionViolatedException e, Description description) {
			logInfo(description, "skipped", nanos);
		}

		@Override
		protected void finished(long nanos, Description description) {
			logInfo(description, "finished", nanos);
		}
	};

	@Test
	public void succeeds() {
	}

	@Test
	// end::user_guide[]
	@Ignore
	// tag::user_guide[]
	public void fails() {
		fail();
	}

	@Test
	public void skips() {
		assumeTrue(false);
	}

	@Test
	public void performanceTest() throws InterruptedException {
		long delta = 30;
		Thread.sleep(300L);
		assertEquals(300d, stopwatch.runtime(MILLISECONDS), delta);
		Thread.sleep(500L);
		assertEquals(800d, stopwatch.runtime(MILLISECONDS), delta);
	}

}
// end::user_guide[]
