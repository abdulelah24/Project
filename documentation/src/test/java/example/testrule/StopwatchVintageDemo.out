tag::user_guide[]
Mar 19, 2017 7:21:25 AM java.util.logging.LogManager$RootLogger log
INFO: Test succeeds succeeded, spent 687 microseconds
Mar 19, 2017 7:21:25 AM java.util.logging.LogManager$RootLogger log
INFO: Test succeeds finished, spent 687 microseconds
Mar 19, 2017 7:21:26 AM java.util.logging.LogManager$RootLogger log
INFO: Test performanceTest succeeded, spent 800549 microseconds
Mar 19, 2017 7:21:26 AM java.util.logging.LogManager$RootLogger log
INFO: Test performanceTest finished, spent 800549 microseconds
Mar 19, 2017 7:21:26 AM java.util.logging.LogManager$RootLogger log
INFO: Test skips skipped, spent 6925 microseconds
Mar 19, 2017 7:21:26 AM java.util.logging.LogManager$RootLogger log
INFO: Test skips finished, spent 6925 microseconds
end::user_guide[]
